homeassistant/components/apple_tv/media_player.py

```diff
@@ -48,49 +49,57 @@
 )
 
 
-async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
-    """Set up the Apple TV platform."""
-    if not discovery_info:
-        return
-
-    # Manage entity cache for service handler
-    if DATA_ENTITIES not in hass.data:
-        hass.data[DATA_ENTITIES] = []
-
-    name = discovery_info[CONF_NAME]
-    host = discovery_info[CONF_HOST]
-    atv = hass.data[DATA_APPLE_TV][host][ATTR_ATV]
-    power = hass.data[DATA_APPLE_TV][host][ATTR_POWER]
-    entity = AppleTvDevice(atv, name, power)
-
-    @callback
-    def on_hass_stop(event):
-        """Stop push updates when hass stops."""
-        atv.push_updater.stop()
-
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, on_hass_stop)
-
-    if entity not in hass.data[DATA_ENTITIES]:
-        hass.data[DATA_ENTITIES].append(entity)
-
-    async_add_entities([entity])
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Load Apple TV media player based on a config entry."""
+    identifier = config_entry.unique_id
+    name = config_entry.data[CONF_NAME]
+    manager = hass.data[DOMAIN][config_entry.unique_id]
+    async_add_entities([AppleTvDevice(name, identifier, manager)])
 
 
 class AppleTvDevice(MediaPlayerDevice):
     """Representation of an Apple TV device."""
 
-    def __init__(self, atv, name, power):
+    def __init__(self, name, identifier, manager):
         """Initialize the Apple TV device."""
-        self.atv = atv
+        self.atv = None
         self._name = name
+        self._identifier = identifier
         self._playing = None
-        self._power = power
-        self._power.listeners.append(self)
-        self.atv.push_updater.listener = self
+        self._manager = manager
 
     async def async_added_to_hass(self):
         """Handle when an entity is about to be added to Home Assistant."""
-        self._power.init()
+        self._manager.listeners.append(self)
+        await self._manager.init()
+
+    async def async_will_remove_from_hass(self):
+        """Handle when an entity is about to be removed from Home Assistant."""
+        self._manager.listeners.remove(self)
+
+    @callback
+    def device_connected(self):
+        """Handle when connection is made to device."""
+        self.atv = self._manager.atv
+        self.atv.push_updater.listener = self
+
+    @callback
+    def device_disconnected(self):
+        """Handle when connection was lost to device."""
+        self.atv.push_updater.listener = None
+        self.atv = None
+
+    @property
+    def device_info(self):
+        """Return the device info."""
+        return {
+            "identifiers": {(DOMAIN, self._identifier)},
+            "manufacturer": "Apple",
+            "model": "Media Player",
+            "name": self.name,
+            "sw_version": "0.0",
+            "via_device": (DOMAIN, self._identifier),
```

Comment:
`via_device` is when you are adding a device that is not the same, but linked to it.

Exemple:
- Philips Hue light via Philips Hue hub
- Synology NAS disk via Synology NAS
- Zwave device via Zwace stick

So not needed here.